Setup dev environment

elhmn opened the pull request:
#### This pull request is adding easy docker-compose setup for local development

**Tasks:**

- [x] Add Dockerfile for the frontend
- [x] Add Dockerfile for the api
- [x] Add gcloud emulator to docker-compose
- [x] Add meilisearch to docker-compose
- [x] Add a python script to setup the datastore emulator
- [x] Add the meillisearch datastore indexer to the docker-compose

**How to run the local environment:**
Run the backend
```sh
docker-compose up api
```

Run the frontend
```sh
docker-compose up frontend
```

**Other commands:**

To teardown running containers and rebuild docker images run: 
```sh
docker-compose up --build --force-recreate frontend
```
To follow docker-compose logs run:
```sh
docker-compose logs --follow
```

run `docker-compose down` to tear down running containers

afranck64 commented on docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+      - 7700:7700
+
+  datastore:
+    image: google/cloud-sdk
```
The google/cloud-sdk image is >~ 1GB ? Can't the alpine or slim version be used instead ?

elhmn replied:
I will create a custom image based on  this file https://github.com/GoogleCloudPlatform/cloud-sdk-docker/blob/master/emulators/Dockerfile

afranck64 commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM node:lts-buster-slim
```
1.  `add #syntax = docker/dockerfile:experimental` is missing at the beginning of the files, which causes mount on line 8 to throw errors.

2. frontend throws errors. Does node:lts-buster-slim point correspond to node14 ?
```
frontend_1           | Failed to compile.
frontend_1           | 
frontend_1           | ./src/styles/index.scss (./node_modules/css-loader/dist/cjs.js??ref--5-oneOf-6-1!./node_modules/postcss-loader/src??postcss!./node_modules/resolve-url-loader??ref--5-oneOf-6-3!./node_modules/sass-loader/dist/cjs.js??ref--5-oneOf-6-4!./src/styles/index.scss)
frontend_1           | Error: Missing binding /code/node_modules/node-sass/vendor/linux-x64-83/binding.node
frontend_1           | Node Sass could not find a binding for your current environment: Linux 64-bit with Node.js 14.x
```

elhmn replied:
I will remove the target cache stuff as they are not very useful in this context. They lead to very weird conflicts, instead we will make use of the Docker layering cache.

afranck64 commented on docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+
+services:
+  #find a way to tear this down after the init scripts has worked
+  setup-datastore:
+    build: ./scripts/setup-datastore/
+    depends_on:
+      - datastore
+    environment:
+      MEILISEARCH_HOST: http://meilisearch:7700
+      GOOGLE_APPLICATION_CREDENTIALS: /tmp/key/creds.json
+      GCLOUD_PROJECT: pname
+      DATASTORE_EMULATOR_HOST: datastore:8000
+      DATASTORE_PROJECT_ID: pname
+
+  #find a way to tear this down after the init scripts has worked
+  setup-meilisearch:
+    build: ./tools/meilisearch-index-generator/
+    depends_on:
+      - setup-datastore
+      - meilisearch
+    environment:
+      MEILISEARCH_HOST: http://meilisearch:7700
+      GOOGLE_APPLICATION_CREDENTIALS: /tmp/key/creds.json
+      GCLOUD_PROJECT: pname
+      DATASTORE_EMULATOR_HOST: datastore:8000
+      DATASTORE_PROJECT_ID: pname
```
setup-datastore_1 throws errors
> setup-datastore_1    | google.api_core.exceptions.ServiceUnavailable: 503 failed to connect to all addresses

Nice to have:
- Enabled cache when building images
- Python package installation while building the image

afranck64 reviewed (changes requested):
If the full build is able to run below 120 secs (or any other arbitrary number considered acceptable that does burn out the Github quotas), it would be nice have it as a test too in the pipeline.

elhmn commented:
> If the full build is able to run below 120 secs (or any other arbitrary number considered acceptable that does burn out the Github quotas), it would be nice have it as a test too in the pipeline.

Yeah you right, but since we are only planning to use this docker-compose locally I don't think it matters too much to have such test to check how long it run for a CI/CD pipeline. Maybe something worth considering in the future if we actually make use of that. I improved the size of the gcloud emulator image a little and added more Docker Layer based caching, and fixed the frontend  dockerfile. @afranck64  do you mind having another look ?
